### ISSUE TITLE
cannot reproduce the results of unsupervised NMT on ende/deen translation task

### ISSUE THREAD

**Issue by tinyka** [facebookresearch/UnsupervisedMT#34]
I can reproduce the results of unsupervised nmt on enfr/fren translation task after 125 epochs:

> epoch -> 125.000000
> ppl_en_fr_valid -> 36.457879
> bleu_en_fr_valid -> 21.250000
> ppl_fr_en_valid -> 49.971469
> bleu_fr_en_valid -> 20.360000
> ppl_en_fr_test -> 20.280857
> **bleu_en_fr_test -> 24.520000**
> ppl_fr_en_test -> 28.081766
> **bleu_fr_en_test -> 24.040000**
> ppl_fr_en_fr_valid -> 1.869078
> bleu_fr_en_fr_valid -> 60.330000
> ppl_fr_en_fr_test -> 1.725349
> bleu_fr_en_fr_test -> 61.530000
> ppl_en_fr_en_valid -> 1.827019
> bleu_en_fr_en_valid -> 62.540000
> ppl_en_fr_en_test -> 1.849352
> bleu_en_fr_en_test -> 62.170000

But I cannot reproduce the results of unsupervised NMT on ende/deen translation task.

I followed the settings in paper and extracted en/de monolingual dataset from wmt14 to wmt17，used newstest2015 as validation set， and newstest2016 as test set. Other settings are same with the settings on enfr translation task.

My command is 
>python main.py --exp_name test --transformer True --n_enc_layers 4 --n_dec_layers 4 --share_enc 3 --share_dec 3 --share_lang_emb True --share_output_emb True --langs 'de,en' --n_mono -1 --mono_dataset 'en:./data_ende/mono/all.en.tok.60000.pth,,;de:./data_ende/mono/all.de.tok.60000.pth,,' --para_dataset 'de-en:,./data_ende/para/dev/newstest2015-ende-ref.XX.60000.pth,./data_ende/para/dev/newstest2016-ende-src.XX.60000.pth' --mono_directions 'de,en' --word_shuffle 3 --word_dropout 0.1 --word_blank 0.2 --pivo_directions 'en-de-en,de-en-de' --pretrained_emb './data_ende/mono/all.en-de.60000.vec' --pretrained_out True --lambda_xe_mono '0:1,100000:0.1,300000:0' --lambda_xe_otfd 1 --otf_num_processes 30 --otf_sync_params_every 1000 --enc_optimizer adam,lr=0.0001 --epoch_size 500000 --stopping_criterion bleu_de_en_valid,100

After 122 epochs,  the bleu on testset is still lower than the reported results:
> epoch -> 122.000000
> ppl_de_en_valid -> 54.630438
> bleu_de_en_valid -> 16.020000
> ppl_en_de_valid -> 57.778759
> bleu_en_de_valid -> 12.210000
> ppl_de_en_test -> 41.127902
> **bleu_de_en_test -> 17.890000**
> ppl_en_de_test -> 42.743846
> **bleu_en_de_test -> 13.520000**
> ppl_de_en_de_valid -> 2.874830
> bleu_de_en_de_valid -> 38.840000
> ppl_de_en_de_test -> 2.799033
> bleu_de_en_de_test -> 38.160000
> ppl_en_de_en_valid -> 2.646685
> bleu_en_de_en_valid -> 46.750000
> ppl_en_de_en_test -> 2.542617
> bleu_en_de_en_test -> 47.100000

I also tried to share all encoder layers and decoder layers following the settings in the paper:
> python main.py --exp_name test --transformer True --n_enc_layers 4 --n_dec_layers 4 --share_enc 4 --share_dec 4 --share_lang_emb True --share_output_emb True --langs 'de,en' --n_mono -1 --mono_dataset 'en:./data_ende/mono/all.en.tok.60000.pth,,;de:./data_ende/mono/all.de.tok.60000.pth,,' --para_dataset 'de-en:,./data_ende/para/dev/newstest2015-ende-ref.XX.60000.pth,./data_ende/para/dev/newstest2016-ende-src.XX.60000.pth' --mono_directions 'de,en' --word_shuffle 3 --word_dropout 0.1 --word_blank 0.2 --pivo_directions 'en-de-en,de-en-de' --pretrained_emb './data_ende/mono/all.en-de.60000.vec' --pretrained_out True --lambda_xe_mono '0:1,100000:0.1,300000:0' --lambda_xe_otfd 1 --otf_num_processes 30 --otf_sync_params_every 1000 --enc_optimizer adam,lr=0.0001 --epoch_size 500000 --stopping_criterion bleu_de_en_valid,100

But got the similar results with only share 3 encoder layers and 3 decoder layers.

Is there anyone could reproduce the results on ende/deen translation task? Is there anything that I miss?

**Comment by glample** [comment]
Hi,

What datasets did you use for English-German? How many sentences do you have?
Can you provide your full training log?

**Comment by glample** [comment]
Hi,

Sorry for the delay on this, I'm not sure what is going on here. I was just able to reproduce the En-De results with my previous code, I didn't try the open source one on the German-English data (I only checked that En-Fr was working properly), I'll give it a try now and get back to you.

**Comment by glample** [comment]
For now, the results seem consistent with what I had before. After epoch 43, on the test set I have: 19.99 BLEU on de->en and 15.64 on en->de.

Regarding your hyper-parameters, I would suggest using 5 layers, but especially sharing all of them:
`--n_enc_layers 5 --n_dec_layers 5 --share_enc 5 --share_dec 5`. The number of layers doesn't seem to matter much, but sharing all of them here seems to make a non-negligible difference.

Freezing the decoder embeddings (`freeze_enc_emb=True`), and not the encoder ones (`freeze_enc_emb=False`) seems to be a bit better than not freezing anything (but overall it doesn't make a big difference).

Maybe also use a `--word_blank 0.1` instead of `0.2`. It's also a negligible difference, but apart from that I basically have the same hyper-parameters.

Also, it takes a few days for the model to get above 21 BLEU, even if the curve is pretty flat around 19 / 20. But in your case clearly the model seems to have converged. I would suggest trying with these hyper-parameters changes above. Also, I took the first 10M lines of:

```
wget -c http://www.statmt.org/wmt14/training-monolingual-news-crawl/news.2007.de.shuffled.gz
wget -c http://www.statmt.org/wmt14/training-monolingual-news-crawl/news.2008.de.shuffled.gz
wget -c http://www.statmt.org/wmt14/training-monolingual-news-crawl/news.2009.de.shuffled.gz
wget -c http://www.statmt.org/wmt14/training-monolingual-news-crawl/news.2010.de.shuffled.gz
```

but the domain should be almost identical (actually your corpora should even have a domain closer to newstest2016) so I don't expect this to make a difference.

**Comment by tinyka** [comment]
@glample Thank you so much for your reply. 
I will rerun the code according to your recommend settings.

**Comment by StillKeepTry** [comment]
@tinyka Did you reproduce the result of ende translation task now?
